nomad/node_endpoint.go

```diff
@@ -1019,10 +1025,25 @@ func (n *Node) UpdateAlloc(args *structs.AllocUpdateRequest, reply *structs.Gene
 
 // batchUpdate is used to update all the allocations
 func (n *Node) batchUpdate(future *structs.BatchFuture, updates []*structs.Allocation, evals []*structs.Evaluation) {
+	// Group pending evals by jobID to prevent creating unnecessary evals
+	evalsByJobId := make(map[structs.NamespacedID]*structs.Evaluation)
```

Comment:
This can be a `map[structs.NamespacedID]struct{}`. It is a bit more space efficient: https://dave.cheney.net/2014/03/25/the-empty-struct